File: nossi/5.Graph/KeysAndRooms.py
# https://leetcode.com/problems/keys-and-rooms/

from collections import deque

def canVisitAllRooms(rooms):

    visited = set()

    def dfs(v):
        visited.add(v)
        for next_v in rooms[v]:
            if next_v not in visited:
                dfs(next_v)
    # dfs(0)

    def bfs(v):
        visited.add(v)
        q = deque()
        q.append(v)
        while q:
            cur_v = q.popleft()
            for next_v in rooms[cur_v]:
                if next_v not in visited:
                    q.append(next_v)
                    visited.add(next_v)

    bfs(0)

    if len(visited) == len(rooms):
        return True
    else:
        return False

# rooms = [[1,3],[2,4],[0],[4],[],[3,4]]
rooms = [[1],[2],[3],[]]
print(canVisitAllRooms(rooms))

File: nossi/5.Graph/KeysAndRooms.js
//  https://leetcode.com/problems/keys-and-rooms/

function canVisitAllRooms(rooms) {
  const visited = new Set();

  function dfs(v) {
    visited.add(v);
    for (let next_v of rooms[v]) {
      if (!visited.has(next_v)) {
        dfs(next_v);
      }
    }
  }

  function bfs(v) {
    visited.add(v);
    const queue = [];
    queue.push(v);
    while (queue.length > 0) {
      const cur_v = queue.shift();
      for (let next_v of rooms[cur_v]) {
        if (!visited.has(next_v)) {
          queue.push(next_v);
          visited.add(next_v);
        }
      }
    }
  }

  //   dfs(0);
  bfs(0);

  console.log(visited);

  if (visited.size === rooms.length) {
    return true;
  } else {
    return false;
  }
}

// rooms = [[1, 3], [2, 4], [0], [4], [], [3, 4]];
rooms = [[1], [2], [3], []];
console.log(canVisitAllRooms(rooms));
